src/bin/wasm-tools/component.rs

```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+            &resolve,
+            world,
+            self.encoding.unwrap_or(StringEncoding::UTF8),
+        )?;
 
-        let bytes = encoder
-            .encode()
-            .with_context(|| format!("failed to encode a component from module ",))?;
+        let section = wasm_encoder::CustomSection {
+            name: "component-type",
+            data: &encoded,
+        };
+        let mut wasm = wasm.unwrap_or_else(|| wit_component::dummy_module(&resolve, world));
+        wasm.push(section.id());
+        section.encode(&mut wasm);
 
         self.io.output(Output::Wasm {
-            bytes: &bytes,
-            wat: self.wat,
+            bytes: &wasm,
+            wat: false,
         })?;
 
         Ok(())
     }
 }
 
-/// WebAssembly interface printer.
+/// Tool for working with the WIT text format for components.
 ///
-/// Decodes a `*.wit` file from a binary WebAssembly component.
+/// This subcommand can be used to inspect and debug the WIT text or binary
+/// format with either WIT packages or binary components. Using this subcommand
+/// a WIT package can be translated to binary, a WIT binary can be translated
+/// back to text, and a WIT document can be extracted from a component binary to
+/// inspect its interfaces.
 #[derive(Parser)]
 pub struct WitOpts {
     #[clap(flatten)]
-    io: wasm_tools::InputOutput,
+    verbosity: wasm_tools::Verbosity,
+
+    /// Input file or directory to process.
+    ///
+    /// The file specified can be a `*.wit` file parsed as a single-document
+    /// package. It can be a directory to be parsed as a WIT package. It can be
+    /// a `*.wat` or `*.wasm` file for either the binary representation of a WIT
+    /// package or a component itself to extract the interface from. The type of
+    /// input is inferred from the contents of the path specified.
+    ///
+    /// If not provided or if this is `-` then stdin is read entirely and
+    /// processed.
+    input: Option<PathBuf>,
 
-    /// The name of the world to generate, inferred by default from the input
-    /// filename.
+    #[clap(flatten)]
+    output: wasm_tools::OutputArg,
+
+    /// If a WIT package is being parsed, then this is the optionally specified
+    /// name of the WIT package. If not specified this is automatically inferred
+    /// from the filename.
     #[clap(long)]
     name: Option<String>,
+
+    /// When printing a WIT package, the default mode, this option is used to
+    /// indicate which document is printed within the package if more than one
+    /// document is present.
+    #[clap(short, long)]
+    document: Option<String>,
+
+    /// Emit a full WIT package into the specified directory when printing the
+    /// text form.
+    ///
+    /// This is incompatible with `-o`.
+    #[clap(long)]
+    out_dir: Option<PathBuf>,
+
+    /// Emit the WIT binary format, a WebAssembly component, instead of the WIT
+    /// text format.
+    #[clap(short, long)]
```

Comment:
I think the `--wasm` and `--wat` options should be mutually exclusive rather than the additive. Requiring `--wat` to need `--wasm` is a little 

Thus the default output is `wit`, `--wasm` generates a wasm binary encoding, and `--wat` generates a wasm text encoding.

```suggestion
    /// Emit a WebAssembly binary representation instead of the WIT text format.
    #[clap(short, long, conflicts_with = "wat")]
```